src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -651,7 +651,33 @@ private BoundDecisionDag MakeBoundDecisionDag(SyntaxNode syntax, ImmutableArray<
 
             var rootDecisionDagNode = decisionDag.RootNode.Dag;
             RoslynDebug.Assert(rootDecisionDagNode != null);
-            return new BoundDecisionDag(rootDecisionDagNode.Syntax, rootDecisionDagNode);
+            var boundDecisionDag = new BoundDecisionDag(rootDecisionDagNode.Syntax, rootDecisionDagNode);
+#if DEBUG
+            int nextTempNumber = 0;
+            // Note that this leverages the `BoundDagEvaluation.GetHashCode()`
+            // to make "equivalent" evaluation nodes share an ID
+            var tempIdentifierMap = PooledDictionary<BoundDagEvaluation, int>.GetInstance();
+            int tempIdentifier(BoundDagEvaluation e)
+            {
+                return tempIdentifierMap.TryGetValue(e, out int value) ? value : tempIdentifierMap[e] = ++nextTempNumber;
+            }
+
+            var sortedBoundDagNodes = boundDecisionDag.TopologicallySortedNodes;
+            for (int i = 0; i < sortedBoundDagNodes.Length; i++)
+            {
+                sortedBoundDagNodes[i].Id = i;
+                if (sortedBoundDagNodes[i] is BoundEvaluationDecisionDagNode { Evaluation: var evaluation })
+                {
+                    evaluation.Id = tempIdentifier(evaluation);
+                    if (evaluation.Input.Source is { } source)
+                    {
+                        source.Id = tempIdentifier(source);
```

Comment:
Why do we need to do this? I would have assumed (by virtue of topological sort) that all sources (ie. previous temps) would have already been assigned an identifier. Feels like we may be assigning identifiers more than once.
Consider adding assertions to prevent double set.

Reply:
Have added a comment. Sometimes the evaluations seem to be the same object instance, other times not, and I'm not sure why, but I do expect these equivalent evaluations to get de-duped during lowering.

Reply:
I think this needs more understanding. We should only have to process nodes forwards. I don't understand why we'd ever have to look back.
What breaks if this is removed?

Reply:
Interesting effects from deleting the `source.Id` assignments here:
- If we delete only the one for BoundDagEvaluation then all the tests actually pass
- If we delete only the one for BoundDagTest then 1 of the tests fails
- If we delete both then 3 of the tests fail

It feels like sometimes the "source" evaluation objects are shared, other times not. Perhaps a matter of convenience during dag construction. I did poke around for a while in binding and lowering of the dags to get a better grasp on this but don't feel 100% on how de-duping works. De-duping seems to sometimes be dependent on the presence/content of when clauses as well.

I would also emphasize that the assumption made here is the same as the assumption made in the original dump code, i.e. it always uses the evaluation object as a dictionary key to look up the id, relying on its equality implementation in order to get the same answer for different instances with equivalent values. I feel like both of the assignments should just be left there since there probably exists a scenario where the absence of just one of them results in uninitialized ids.

One scenario which breaks when we remove these assignments: [DecisionDag_Dump_SwitchStatement_01](https://github.com/RikkiGibson/roslyn/blob/35ebde268069bacab998ff022104d8eeac05a7e3/src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests4.cs#L3511-L3511)
```diff
[1]: t1 = (string)t0; [2]
 [2]: t1 == ""a"" ? [3] : [4]
 [3]: leaf `case ""a"":`
-[4]: t2 = t1.Length; [5]
+[4]: t2 = <uninitialized>.Length; [5]
 [5]: t2 == 1 ? [6] : [13]
 [6]: when <true> ? [7] : <unreachable>
 [7]: leaf `case string { Length: 1 } s:`
```

Reply:
I'll give this a try locally.

Reply:
I stepped through `DecisionDag_Dump_SwitchStatement_01` and I think your change is right.